src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1269,5 +1226,270 @@ internal override bool CanAddProjectReference(ProjectId referencingProject, Proj
             var result = referenceManager.QueryCanReferenceProject(referencingHierarchy, referencedHierarchy);
             return result != (uint)__VSREFERENCEQUERYRESULT.REFERENCE_DENY;
         }
+
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// Applies a single operation to the workspace. <paramref name="action"/> should be a call to one of the protected Workspace.On* methods.
+        /// </summary>
+        public void ApplyChangeToWorkspace(Action<Workspace> action)
+        {
+            lock (_gate)
+            {
+                action(this);
+            }
+        }
+
+        /// <summary>
+        /// Applies a change to the workspace that can do any number of project changes.
+        /// </summary>
+        /// <remarks>This is needed to synchronize with <see cref="ApplyChangeToWorkspace(Action{Workspace})" /> to avoid any races. This
+        /// method could be moved down to the core Workspace layer and then could use the synchronization lock there.</remarks>
+        /// <param name="projectId">The <see cref="ProjectId" /> to change.</param>
+        /// <param name="mutation">A function that, given the old <see cref="CodeAnalysis.Project"/> will produce a new one.</param>
+        public void ApplyBatchChangeToProject(ProjectId projectId, Func<CodeAnalysis.Solution, CodeAnalysis.Solution> mutation)
+        {
+            lock (_gate)
+            {
+                var oldSolution = this.CurrentSolution;
+                var newSolution = mutation(oldSolution);
+
+                if (oldSolution == newSolution)
+                {
+                    return;
+                }
+
+                SetCurrentSolution(newSolution);
+                RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectChanged, oldSolution, newSolution, projectId);
+            }
+        }
+
+        private Dictionary<ProjectId, ProjectReferenceInformation> _projectReferenceInfo = new Dictionary<ProjectId, ProjectReferenceInformation>();
```

Comment:
> private Dictionary<ProjectId, ProjectReferenceInformation> _projectReferenceInfo = new Dictionary<ProjectId, ProjectReferenceInformation>(); [](start = 7, length = 141)

Move to the top.

Reply:
Going to leave these here for now -- in some sense we may want to move this over to a separate partial...